File: src/main/java/com/hms/patient/application/VisitService.java
package com.hms.patient.application;

import com.hms.patient.application.dto.request.CreateVisitRequest;
import com.hms.patient.application.dto.response.VisitResponse;
import com.hms.patient.domain.model.entity.PatientVisit.VisitStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service interface for patient visit (OPD) operations.
 */
public interface VisitService {

    /**
     * Creates a new patient visit.
     *
     * @param request the visit creation request
     * @param createdBy the user ID of the creator
     * @return the created visit response
     */
    VisitResponse createVisit(CreateVisitRequest request, Long createdBy);

    /**
     * Updates visit status.
     *
     * @param uuid the visit UUID
     * @param status the new status
     * @param updatedBy the user ID of the updater
     * @return the updated visit response
     */
    VisitResponse updateVisitStatus(UUID uuid, VisitStatus status, Long updatedBy);

    /**
     * Gets a visit by UUID.
     *
     * @param uuid the visit UUID
     * @return the visit response
     */
    Optional<VisitResponse> getVisitByUuid(UUID uuid);

    /**
     * Gets a visit by visit number.
     *
     * @param visitNumber the visit number
     * @return the visit response
     */
    Optional<VisitResponse> getVisitByNumber(String visitNumber);

    /**
     * Gets visits by patient UUID.
     *
     * @param patientUuid the patient UUID
     * @return list of visit responses
     */
    List<VisitResponse> getVisitsByPatientUuid(UUID patientUuid);

    /**
     * Gets visits by patient ID with pagination.
     *
     * @param patientId the patient ID
     * @param pageable pagination information
     * @return page of visit responses
     */
    Page<VisitResponse> getVisitsByPatientId(Long patientId, Pageable pageable);

    /**
     * Gets today's visits for a department.
     *
     * @param departmentId the department ID
     * @param pageable pagination information
     * @return page of visit responses
     */
    Page<VisitResponse> getTodayVisitsByDepartment(Long departmentId, Pageable pageable);

    /**
     * Gets the queue for a department.
     *
     * @param departmentId the department ID
     * @return list of visit responses in queue order
     */
    List<VisitResponse> getDepartmentQueue(Long departmentId);

    /**
     * Gets visits by doctor and status.
     *
     * @param doctorId the doctor ID
     * @param status the visit status
     * @return list of visit responses
     */
    List<VisitResponse> getVisitsByDoctorAndStatus(Long doctorId, VisitStatus status);

    /**
     * Gets the next token number for a department.
     *
     * @param departmentId the department ID
     * @return the next token number
     */
    Integer getNextTokenNumber(Long departmentId);

    /**
     * Checks in a patient for their visit.
     *
     * @param uuid the visit UUID
     * @return the updated visit response
     */
    VisitResponse checkInPatient(UUID uuid);

    /**
     * Gets all visits with pagination.
     *
     * @param pageable pagination information
     * @return page of visit responses
     */
    Page<VisitResponse> getAllVisits(Pageable pageable);

    /**
     * Gets visits by status.
     *
     * @param status the visit status
     * @return list of visit responses
     */
    List<VisitResponse> getVisitsByStatus(VisitStatus status);
}

File: src/app/core/services/visit.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '@env/environment';
import { Visit, CreateVisitRequest, UpdateVisitRequest, VisitStatus } from '../models/visit.model';
import { PageableResponse } from './patient.service';

@Injectable({
  providedIn: 'root'
})
export class VisitService {
  private http = inject(HttpClient);
  private baseUrl = `${environment.apiUrl}/visits`;

  createVisit(request: CreateVisitRequest): Observable<Visit> {
    return this.http.post<Visit>(this.baseUrl, request);
  }

  updateVisit(uuid: string, request: UpdateVisitRequest): Observable<Visit> {
    return this.http.put<Visit>(`${this.baseUrl}/${uuid}`, request);
  }

  getVisit(uuid: string): Observable<Visit> {
    return this.http.get<Visit>(`${this.baseUrl}/${uuid}`);
  }

  getVisitByNumber(visitNumber: string): Observable<Visit> {
    return this.http.get<Visit>(`${this.baseUrl}/number/${visitNumber}`);
  }

  getPatientVisits(patientUuid: string): Observable<Visit[]> {
    return this.http.get<Visit[]>(`${this.baseUrl}/patient/${patientUuid}/visits`);
  }

  getPatientVisitsById(patientId: number, page: number = 0, size: number = 20): Observable<PageableResponse<Visit>> {
    return this.http.get<PageableResponse<Visit>>(`${this.baseUrl}/patient/${patientId}?page=${page}&size=${size}`);
  }

  getTodayVisitsByDepartment(departmentId: number, page: number = 0, size: number = 50): Observable<PageableResponse<Visit>> {
    return this.http.get<PageableResponse<Visit>>(`${this.baseUrl}/department/${departmentId}/today?page=${page}&size=${size}`);
  }

  getDepartmentQueue(departmentId: number): Observable<Visit[]> {
    return this.http.get<Visit[]>(`${this.baseUrl}/department/${departmentId}/queue`);
  }

  getAllVisits(): Observable<Visit[]> {
    return this.http.get<{ content: Visit[] }>(this.baseUrl).pipe(
      map(response => response.content)
    );
  }

  getVisitsByStatus(status: VisitStatus): Observable<Visit[]> {
    return this.http.get<Visit[]>(`${this.baseUrl}/status/${status}`);
  }

  getDoctorVisits(doctorId: number, status: VisitStatus = VisitStatus.IN_QUEUE): Observable<Visit[]> {
    return this.http.get<Visit[]>(`${this.baseUrl}/doctor/${doctorId}/visits?status=${status}`);
  }

  updateVisitStatus(uuid: string, status: VisitStatus): Observable<Visit> {
    return this.http.patch<Visit>(`${this.baseUrl}/${uuid}/status?status=${status}`, {});
  }

  checkInPatient(uuid: string): Observable<Visit> {
    return this.http.post<Visit>(`${this.baseUrl}/${uuid}/check-in`, {});
  }

  saveConsultationNotes(
    uuid: string,
    diagnosis: string,
    treatmentPlan?: string,
    prescription?: string
  ): Observable<Visit> {
    let params = new HttpParams().set('diagnosis', diagnosis);
    if (treatmentPlan) params = params.set('treatmentPlan', treatmentPlan);
    if (prescription) params = params.set('prescription', prescription);
    return this.http.post<Visit>(`${this.baseUrl}/${uuid}/consultation`, {}, { params });
  }

  referPatient(
    uuid: string,
    referredDepartmentId: number,
    referredDepartmentName?: string,
    referralReason?: string
  ): Observable<Visit> {
    let params = new HttpParams().set('referredDepartmentId', referredDepartmentId.toString());
    if (referredDepartmentName) params = params.set('referredDepartmentName', referredDepartmentName);
    if (referralReason) params = params.set('referralReason', referralReason);
    return this.http.post<Visit>(`${this.baseUrl}/${uuid}/refer`, {}, { params });
  }
}
